WT-15527 When picking up a checkpoint, open shared metadata at that checkpoint.

ddanderson opened the pull request:
This has been tested by running tests that pick up checkpoints (e.g. test_layered23), stopping in the debugger and verifying that the shared URI is correct.

Commit:
WT-15527 When picking up a checkpoint, open shared metadata at that checkpoint.

quchenhao commented on src/conn/conn_layered.c:
Can we also remove __wti_conn_dhandle_outdated? I think for a checkpoint dhandle, it should be reclaimed by the sweep server automatically? I'm not sure how this is supposed to work though, I may be wrong and this is still needed.

pmacko86 commented on src/conn/conn_layered.c:
```diff
@@ -507,9 +508,23 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+     * "(WiredTigerCheckpoint.NNN=(...))", so extract the checkpoint name first.
+     */
+    checkpoint_name = metadata_value;
+    WT_PREFIX_SKIP_REQUIRED(session, checkpoint_name, "(");
```
As this has to do with validation of external inputs, please use constructs that would be present in the release build, and log an error on mismatch (e.g., via `WT_ERR_MSG`).

pmacko86 commented on src/conn/conn_layered.c:
```diff
@@ -507,9 +508,23 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
+    checkpoint_name = metadata_value;
+    WT_PREFIX_SKIP_REQUIRED(session, checkpoint_name, "(");
+    end = strchr(checkpoint_name, '=');
+    WT_ASSERT(session, end != NULL);
```
Same here.

pmacko86 reviewed (approved):
LGTM after addressing the input validation issues.

Commit:
Change macro so that a metadata at a checkpoint is still considered
metadata.  Fix asserts in metadata config checking to be errors.
Remove call to "outdate" the old metadata file.

quchenhao reviewed (approved):
lgtm if the tests are fixed

ddanderson commented:
I'm turning this PR into a draft for now, I'm not actively pursuing it (see discussion on the ticket).

radoslavkardum commented:
Hi Don, we should close this PR as we've resolved the ticket and merged the changes as part of #13411 PR

ddanderson commented:
Closing this PR, it's fixed in #13411
